File: app/src/main/java/ado/com/testarchcomponents/viewmodel/LocationViewModel.java
package ado.com.testarchcomponents.viewmodel;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.support.annotation.NonNull;

import ado.com.testarchcomponents.model.GpsLocation;

public class LocationViewModel extends AndroidViewModel {

    private final LiveData<GpsLocation> mLocationLiveData;

    public LocationViewModel(@NonNull Application application) {
        super(application);
        mLocationLiveData = new LocationLiveData(new LocationManager(application));
    }

    public LiveData<GpsLocation> getLocationLiveData() {
        return mLocationLiveData;
    }


    private class LocationLiveData extends LiveData<GpsLocation> {
        private final LocationManager mListener2;

        private GpsLocationListener mGpsLocationListener = this::setValue;

        private LocationLiveData(final LocationManager listener) {
            mListener2 = listener;
        }

        @Override
        protected void onActive() {
            mListener2.start(mGpsLocationListener);
        }

        @Override
        protected void onInactive() {
            mListener2.stop();
        }
    }
}

File: app/src/main/java/ado/com/testarchcomponents/viewmodel/LocationViewModel.kt
package ado.com.testarchcomponents.viewmodel

import android.app.Application
import android.arch.lifecycle.AndroidViewModel
import android.arch.lifecycle.LiveData

import ado.com.testarchcomponents.model.GpsLocation

class LocationViewModel(application: Application) : AndroidViewModel(application) {

    val locationLiveData: LiveData<GpsLocation>

    init {
        locationLiveData = LocationLiveData(GpsLocationManager(application))
    }

    private class LocationLiveData
    constructor(private val mGpsLocationManager: GpsLocationManager) : LiveData<GpsLocation>() {

        //private val mGpsLocationListener = GpsLocationListener { this.value = it }

        private val mGpsLocationListener = object : GpsLocationListener {
            override fun onLocationChange(location: GpsLocation) { value = location }
        }

        override fun onActive() {
            mGpsLocationManager.start(mGpsLocationListener)
        }

        override fun onInactive() {
            mGpsLocationManager.stop()
        }
    }
}
